network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -537,90 +543,250 @@ def pretty_print_diff(diff_list):
-                removed[ConnectionSet.protocol_number_to_name(protocol)] = {}
-            added['All connections'] = []
-            removed['All connections'] = []
-            # Hash diffs: protocol-> port ranges-> list of (from endpoint, to endpoint) tuples
-            for entry in all_diffs:
-                if entry.added.allow_all:
-                    added['All connections'].append((entry.from_ep, entry.to_ep))
-                else:
-                    for protocol in entry.added.allowed_protocols:
-                        if not ConnectionSet.protocol_supports_ports(protocol):
-                            continue
-                        protocol_name = ConnectionSet.protocol_number_to_name(protocol)
-                        port_range = str(entry.added.allowed_protocols[protocol])
-                        if port_range not in added[protocol_name]:
-                            added[protocol_name][port_range] = []
-                        added[protocol_name][port_range].append((entry.from_ep, entry.to_ep))
-                if entry.removed.allow_all:
-                    removed['All connections'].append((entry.from_ep, entry.to_ep))
-                else:
-                    for protocol in entry.removed.allowed_protocols:
-                        if not ConnectionSet.protocol_supports_ports(protocol):
-                            continue
-                        protocol_name = ConnectionSet.protocol_number_to_name(protocol)
-                        port_range = str(entry.removed.allowed_protocols[protocol])
-                        if port_range not in removed[protocol_name]:
-                            removed[protocol_name][port_range] = []
-                        removed[protocol_name][port_range].append((entry.from_ep, entry.to_ep))
-            explanation = f'Added connections:\n{self.pretty_print_diff(added)}\n'
-            explanation += f'Removed connections:\n{self.pretty_print_diff(removed)}\n'
-
-            return QueryAnswer(False, 'Base branch and analyzed commit are not semantically equivalent.', explanation)
-        return QueryAnswer(True, 'Base branch and analyzed commit are semantically equivalent.')
+                    all_diff[key].append(
+                        SemanticDiffQuery.SingleDiff(pod1, pod2, old_conns - new_conns, new_conns - old_conns))
+
+        """
+        3.2  lost connections between intersected peers and removed ipBlock peers
+        3.3  lost/new connections between intersected peers and intersected ipBlock peers due to changes in policies and labels of pods/namespaces
+        3.4  new connections between intersected peers and added ipBlock peers
+        """
+        key = 'Changed connections between persistent peers and ipBlock peers'
+        all_diff[key] = []
+        peers = captured_pods | disjoint_ip_blocks
+        for pod1 in peers:
+            for pod2 in disjoint_ip_blocks if pod1 in captured_pods else captured_pods:
```

Comment:
This is wrong.
It will only catch changes in "pod to ipblock" connections.
It will not catch changes in "ipblock to pod" connections.

Reply:
one thing I was missing here is relating pods that are either captured in old-config and uncaptured in new-config or vice versa. for such pods we need to compute also against the unreferenced ip-blocks.
fixed. pls take a close look as this is the most tricky part :)
also enhanced the test to cover for this use-case and checked for correctness of the output.